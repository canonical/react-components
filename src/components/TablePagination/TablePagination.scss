@import "vanilla-framework/scss/settings";

.pagination {
  align-items: baseline;
  display: flex;
  margin: $spv--large 0;

  @media screen and (max-width: $breakpoint-small) {
    justify-content: flex-end;
  }

  .description {
    flex-grow: 1;

    @media screen and (max-width: $breakpoint-small) {
      display: none;
    }
  }

  .back {
    margin: 0 0 0 $spv--large;

    .p-icon--chevron-down {
      rotate: 90deg;
    }
  }

  .next {
    margin: 0 $spv--large;

    .p-icon--chevron-down {
      rotate: 270deg;
    }
  }

  .pagination-input {
    margin: 0 $spv--small 0 $spv--large;
    min-width: 0;
    width: 3rem;
  }

  .pagination-select {
    margin-bottom: 0;
    margin-left: $spv--x-large;
    min-width: 0;
    width: 7rem;
  }
}
