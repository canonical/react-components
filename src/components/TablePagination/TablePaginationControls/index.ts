export { default } from "./TablePaginationControls";
export type { Props as TablePaginationControlsProps } from "./TablePaginationControls";
