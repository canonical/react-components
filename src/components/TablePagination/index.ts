export { default } from "./TablePagination";
export type { Props as TablePaginationProps } from "./TablePagination";
