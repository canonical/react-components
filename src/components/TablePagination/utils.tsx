import {
  Children,
  ReactElement,
  ReactNode,
  cloneElement,
  useEffect,
  useState,
} from "react";

/**
 * Determine if we are working with a small screen.
 * 'small screen' in this case is relative to the width of the description div
 */
export const figureSmallScreen = () => {
  const descriptionElement = document.getElementById("pagination-description");
  if (!descriptionElement) {
    return true;
  }
  return descriptionElement.getBoundingClientRect().width < 230;
};

/**
 * Iterate direct react child components and override the value of the prop specified by @param dataForwardProp
 * for those child components.
 * @param children - react node children to iterate
 * @param dataForwardProp - the name of the prop from the children components to override
 * @param data - actual data to be passed to the prop specified by @param dataForwardProp
 */
export const renderChildren = (
  children: ReactNode,
  dataForwardProp: string,
  data: unknown[]
) => {
  return Children.map(children, (child) => {
    return cloneElement(child as ReactElement, {
      [dataForwardProp]: data,
    });
  });
};

export const DEFAULT_PAGE_LIMITS = [50, 100, 200];
export const generatePagingOptions = (pageLimits: number[]) => {
  return pageLimits.map((limit) => ({ value: limit, label: `${limit}/page` }));
};

export const getDescription = ({
  description,
  data,
  isSmallScreen,
  totalItems,
  itemName,
}: {
  description: ReactNode;
  data: unknown[];
  isSmallScreen: boolean;
  totalItems: number;
  itemName: string;
}) => {
  if (description) {
    return description;
  }

  const visibleCount = data.length;

  if (isSmallScreen) {
    return `${visibleCount} out of ${totalItems}`;
  }

  if (visibleCount === totalItems && visibleCount > 1) {
    return `Showing all ${totalItems} ${itemName}s`;
  }

  return `Showing ${visibleCount} out of ${totalItems} ${itemName}${
    totalItems !== 1 ? "s" : ""
  }`;
};

export const useFigureSmallScreen = () => {
  const [isSmallScreen, setSmallScreen] = useState(false);
  useEffect(() => {
    const handleResize = () => {
      setSmallScreen(figureSmallScreen());
    };
    window.addEventListener("resize", handleResize);
    return () => {
      window.removeEventListener("resize", handleResize);
    };
  }, []);

  return isSmallScreen;
};
