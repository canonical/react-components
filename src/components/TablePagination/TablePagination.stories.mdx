import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";

import TablePagination from "./TablePagination";
import MainTable from "../MainTable";

<Meta title="TablePagination" component={TablePagination} />

export const Template = (args) => <TablePagination {...args} />;

### TablePagination

This is an HOC [React](https://reactjs.org/) component for applying pagination to input data for direct child components.
This component is un-opinionated about the structure of the input data and can be used with any child component that displays 
a list of data. However, the styling and behaviour of this component were designed to work nicely with the ```MainTable``` component.

To use this component, simply wrap a child component with it and provide the data that you want to paginate to the ```data``` prop. 
This component will then pass the paged data to all <b>direct</b> child components via a child prop specified by ```dataForwardProp```.  

### Props

<ArgsTable of={TablePagination} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{
      data: [{id: "row-1"}, {id: "row-2"}, {id: "row-3"}, {id: "row-4"}, {id: "row-5"}],
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Custom page limit

<Canvas>
  <Story
    name="CustomPageLimit"
    args={{
      data: [{id: "row-1"}, {id: "row-2"}, {id: "row-3"}, {id: "row-4"}, {id: "row-5"}],
      pageLimits: [1, 2, 3]
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Custom display title

<Canvas>
  <Story
    name="CustomDisplayTitle"
    args={{
      data: [{id: "row-1"}, {id: "row-2"}, {id: "row-3"}, {id: "row-4"}, {id: "row-5"}],
      description: <b>Hello there</b>
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Render above

<Canvas>
  <Story
    name="RenderAbove"
  >
    {() => {
      const data = [
        {
          columns: [
            { content: "Ready", role: "rowheader" },
            { content: 1, className: "u-align--right" },
            { content: "1 GiB", className: "u-align--right" },
            { content: 2, className: "u-align--right" },
            { content: 42, className: "u-align--right" },
          ],
          sortData: {
            status: "ready",
            cores: 2,
            ram: 1,
            disks: 2,
          },
        },
        {
          columns: [
            { content: "Idle", role: "rowheader" },
            { content: 1, className: "u-align--right" },
            { content: "1 GiB", className: "u-align--right" },
            { content: 2, className: "u-align--right" },
            { content: 23, className: "u-align--right" },
          ],
          sortData: {
            status: "idle",
            cores: 1,
            ram: 1,
            disks: 2,
          },
        },
        {
          columns: [
            { content: "Waiting", role: "rowheader" },
            { content: 8, className: "u-align--right" },
            { content: "3.9 GiB", className: "u-align--right" },
            { content: 3, className: "u-align--right" },
            { content: 0, className: "u-align--right" },
          ],
          sortData: {
            status: "waiting",
            cores: 8,
            ram: 3.9,
            disks: 3,
          },
        },
      ];

      const headers = [
        { content: "Status", sortKey: "status" },
        { content: "Cores", sortKey: "cores", className: "u-align--right" },
        { content: "RAM", sortKey: "ram", className: "u-align--right" },
        { content: "Disks", sortKey: "disks", className: "u-align--right" },
        { content: "Networks", className: "u-align--right" },
      ];

      return (
        <TablePagination data={data} pageLimits={[1, 2, 3]}>
          <MainTable
            headers={headers}
            rows={data}
            sortabl
          />
        </TablePagination>
      );
    }}
  </Story>
</Canvas>

### Render below

<Canvas>
  <Story
    name="RenderBelow"
  >
    {() => {
      const data = [
        {
          columns: [
            { content: "Ready", role: "rowheader" },
            { content: 1, className: "u-align--right" },
            { content: "1 GiB", className: "u-align--right" },
            { content: 2, className: "u-align--right" },
            { content: 42, className: "u-align--right" },
          ],
          sortData: {
            status: "ready",
            cores: 2,
            ram: 1,
            disks: 2,
          },
        },
        {
          columns: [
            { content: "Idle", role: "rowheader" },
            { content: 1, className: "u-align--right" },
            { content: "1 GiB", className: "u-align--right" },
            { content: 2, className: "u-align--right" },
            { content: 23, className: "u-align--right" },
          ],
          sortData: {
            status: "idle",
            cores: 1,
            ram: 1,
            disks: 2,
          },
        },
        {
          columns: [
            { content: "Waiting", role: "rowheader" },
            { content: 8, className: "u-align--right" },
            { content: "3.9 GiB", className: "u-align--right" },
            { content: 3, className: "u-align--right" },
            { content: 0, className: "u-align--right" },
          ],
          sortData: {
            status: "waiting",
            cores: 8,
            ram: 3.9,
            disks: 3,
          },
        },
      ];

      const headers = [
        { content: "Status", sortKey: "status" },
        { content: "Cores", sortKey: "cores", className: "u-align--right" },
        { content: "RAM", sortKey: "ram", className: "u-align--right" },
        { content: "Disks", sortKey: "disks", className: "u-align--right" },
        { content: "Networks", className: "u-align--right" },
      ];

      return (
        <TablePagination data={data} pageLimits={[1, 2, 3]} position="below">
          <MainTable
            headers={headers}
            rows={data}
            sortabl
          />
        </TablePagination>
      );
    }}
  </Story>
</Canvas>
