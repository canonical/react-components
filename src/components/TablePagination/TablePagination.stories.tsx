import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import TablePagination from "./TablePagination";
import MainTable from "../MainTable";

const meta: Meta<typeof TablePagination> = {
  component: TablePagination,
  tags: ["autodocs"],

  argTypes: {
    totalItems: {
      if: {
        arg: "externallyControlled",
        truthy: true,
      },
    },

    currentPage: {
      if: {
        arg: "externallyControlled",
        truthy: true,
      },
    },

    pageSize: {
      if: {
        arg: "externallyControlled",
        truthy: true,
      },
    },

    onPageChange: {
      if: {
        arg: "externallyControlled",
        truthy: true,
      },
    },

    onPageSizeChange: {
      if: {
        arg: "externallyControlled",
        truthy: true,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof TablePagination>;

export const Default: Story = {
  name: "Default",

  args: {
    data: [
      {
        id: "row-1",
      },
      {
        id: "row-2",
      },
      {
        id: "row-3",
      },
      {
        id: "row-4",
      },
      {
        id: "row-5",
      },
    ],
  },
};

export const CustomPageLimit: Story = {
  name: "CustomPageLimit",

  args: {
    data: [
      {
        id: "row-1",
      },
      {
        id: "row-2",
      },
      {
        id: "row-3",
      },
      {
        id: "row-4",
      },
      {
        id: "row-5",
      },
    ],

    pageLimits: [1, 2, 3],
  },
};

export const CustomDisplayTitle: Story = {
  name: "CustomDisplayTitle",

  args: {
    data: [
      {
        id: "row-1",
      },
      {
        id: "row-2",
      },
      {
        id: "row-3",
      },
      {
        id: "row-4",
      },
      {
        id: "row-5",
      },
    ],

    description: <b>Hello there</b>,
  },
};

export const RenderAbove: Story = {
  render: () => {
    const data = [
      {
        columns: [
          {
            content: "Ready",
            role: "rowheader",
          },
          {
            content: 1,
            className: "u-align--right",
          },
          {
            content: "1 GiB",
            className: "u-align--right",
          },
          {
            content: 2,
            className: "u-align--right",
          },
          {
            content: 42,
            className: "u-align--right",
          },
        ],

        sortData: {
          status: "ready",
          cores: 2,
          ram: 1,
          disks: 2,
        },
      },
      {
        columns: [
          {
            content: "Idle",
            role: "rowheader",
          },
          {
            content: 1,
            className: "u-align--right",
          },
          {
            content: "1 GiB",
            className: "u-align--right",
          },
          {
            content: 2,
            className: "u-align--right",
          },
          {
            content: 23,
            className: "u-align--right",
          },
        ],

        sortData: {
          status: "idle",
          cores: 1,
          ram: 1,
          disks: 2,
        },
      },
      {
        columns: [
          {
            content: "Waiting",
            role: "rowheader",
          },
          {
            content: 8,
            className: "u-align--right",
          },
          {
            content: "3.9 GiB",
            className: "u-align--right",
          },
          {
            content: 3,
            className: "u-align--right",
          },
          {
            content: 0,
            className: "u-align--right",
          },
        ],

        sortData: {
          status: "waiting",
          cores: 8,
          ram: 3.9,
          disks: 3,
        },
      },
    ];

    const headers = [
      {
        content: "Status",
        sortKey: "status",
      },
      {
        content: "Cores",
        sortKey: "cores",
        className: "u-align--right",
      },
      {
        content: "RAM",
        sortKey: "ram",
        className: "u-align--right",
      },
      {
        content: "Disks",
        sortKey: "disks",
        className: "u-align--right",
      },
      {
        content: "Networks",
        className: "u-align--right",
      },
    ];

    return (
      <TablePagination data={data} pageLimits={[1, 2, 3]}>
        <MainTable headers={headers} rows={data} sortable />
      </TablePagination>
    );
  },

  name: "RenderAbove",
};

export const RenderBelow: Story = {
  render: () => {
    const data = [
      {
        columns: [
          {
            content: "Ready",
            role: "rowheader",
          },
          {
            content: 1,
            className: "u-align--right",
          },
          {
            content: "1 GiB",
            className: "u-align--right",
          },
          {
            content: 2,
            className: "u-align--right",
          },
          {
            content: 42,
            className: "u-align--right",
          },
        ],

        sortData: {
          status: "ready",
          cores: 2,
          ram: 1,
          disks: 2,
        },
      },
      {
        columns: [
          {
            content: "Idle",
            role: "rowheader",
          },
          {
            content: 1,
            className: "u-align--right",
          },
          {
            content: "1 GiB",
            className: "u-align--right",
          },
          {
            content: 2,
            className: "u-align--right",
          },
          {
            content: 23,
            className: "u-align--right",
          },
        ],

        sortData: {
          status: "idle",
          cores: 1,
          ram: 1,
          disks: 2,
        },
      },
      {
        columns: [
          {
            content: "Waiting",
            role: "rowheader",
          },
          {
            content: 8,
            className: "u-align--right",
          },
          {
            content: "3.9 GiB",
            className: "u-align--right",
          },
          {
            content: 3,
            className: "u-align--right",
          },
          {
            content: 0,
            className: "u-align--right",
          },
        ],

        sortData: {
          status: "waiting",
          cores: 8,
          ram: 3.9,
          disks: 3,
        },
      },
    ];

    const headers = [
      {
        content: "Status",
        sortKey: "status",
      },
      {
        content: "Cores",
        sortKey: "cores",
        className: "u-align--right",
      },
      {
        content: "RAM",
        sortKey: "ram",
        className: "u-align--right",
      },
      {
        content: "Disks",
        sortKey: "disks",
        className: "u-align--right",
      },
      {
        content: "Networks",
        className: "u-align--right",
      },
    ];

    return (
      <TablePagination data={data} pageLimits={[1, 2, 3]} position="below">
        <MainTable headers={headers} rows={data} sortable />
      </TablePagination>
    );
  },

  name: "RenderBelow",
};
