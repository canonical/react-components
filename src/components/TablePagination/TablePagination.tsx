import React, {
  HTMLAttributes,
  PropsWithChildren,
  ReactNode,
  useState,
} from "react";
import "./TablePagination.scss";
import TablePaginationControls from "./TablePaginationControls";
import {
  DEFAULT_PAGE_LIMITS,
  generatePagingOptions,
  renderChildren,
} from "./utils";
import { usePagination } from "hooks";

export type BasePaginationProps = {
  /**
   * list of data elements to be paginated. This component is un-opinionated about
   * the structure of the data but it should be identical to the data structure
   * reuiqred by the child table component
   */
  data: unknown[];
  /**
   * prop name of the child table component that receives paginated data.
   * default value is set to `rows`, which is the data prop for the `MainTable` component
   */
  dataForwardProp?: string;
  /**
   * the name of the item associated to each row within the table.
   */
  itemName?: string;
  /**
   * custom styling for the pagination container
   */
  className?: string;
  /**
   * custom description to be displayed by the pagination
   */
  description?: ReactNode;
  /**
   * custom per page limits express as an array of numbers.
   */
  pageLimits?: number[];
  /**
   * place the pagination component above or below the table?
   */
  position?: "above" | "below";
};

export type ExternalControlProps = BasePaginationProps & {
  /**
   * Whether the component will be controlled via external state.
   */
  externallyControlled?: true;
  /**
   * the total number of items available within the data. This prop is only relevant
   * and will be required if `externallyControlled` is set to `true`.
   */
  totalItems: number;
  /**
   * the current page that's showing. This prop is only relevant and will be required
   * if `externallyControlled` is set to `true`.
   */
  currentPage: number;
  /**
   * size per page. This prop is only relevant and will be required if
   * `externallyControlled` is set to `true`.
   */
  pageSize: number;
  /**
   * callback indicating a page change event to the parent component.
   * This prop is only relevant and will be required if `externallyControlled` is set
   * to `true`.
   */
  onPageChange: (page: number) => void;
  /**
   * callback indicating a page size change event to the parent component.
   * This prop is only relevant and will be required if `externallyControlled` is set
   * to `true`.
   */
  onPageSizeChange: (pageSize: number) => void;
};

export type InternalControlProps = BasePaginationProps & {
  /**
   * Whether the component will be controlled via external state.
   */
  externallyControlled?: false;
};

export type Props = PropsWithChildren<
  ExternalControlProps | InternalControlProps
> &
  HTMLAttributes<HTMLDivElement>;

/**
This is an HOC [React](https://reactjs.org/) component for applying pagination to direct children components. This component is un-opinionated about
the structure of the input data and can be used with any child component that displays a list of data. However, the styling and behaviour of this component were designed
to work nicely with the `MainTable` component. To use this component, simply wrap a child component with it and provide the data that you want
to paginate to the `data` prop. This component will then pass the paged data to all <b>direct</b> child components via a child prop specified by `dataForwardProp`.
The component may be externally controlled, see following sections for detailed explanation.

#### Externally controlled pagination

For externally controlled mode, you will be responsible for the pagination logic and therefore the component will be purely presentational.
The pagination behaviour is controlled outside of this component. Note the data injection to child components is essentially a passthrough in this case.
To enable externally controlled mode for this component, set the `externallyControlled` prop to `true`. From there, it is your responsibility
to ensure that the following props `totalItems`, `currentPage`, `pageSize`, `onPageChange` and `onPageSizeChange` are set properly.
You can refer to the props table below on how to set these props.

#### Un-controlled pagination

In this mode, the component assumes that the input data is not paginated. The component will implement the pagination logic and apply it to the input data
then inject the paged data into direct child components. This is the default mode of operations for the component where `externallyControlled` prop is set
to `false`.
 */
const TablePagination = (props: Props) => {
  const {
    data,
    dataForwardProp = "rows",
    itemName = "item",
    className,
    description,
    pageLimits = DEFAULT_PAGE_LIMITS,
    position = "above",
    externallyControlled,
    children,
    ...divProps
  } = props;

  // Safety check to ensure pageSize is a valid option in
  // pageLimits if the component is externally controlled
  if (externallyControlled) {
    let pageSizeFound = false;
    for (const limit of pageLimits) {
      if (limit === Number(props.pageSize)) {
        pageSizeFound = true;
        break;
      }
    }

    if (!pageSizeFound) {
      throw new Error(
        `pageSize must be a valid option in pageLimits, pageLimits is set to [${pageLimits}]`
      );
    }
  }

  const [internalPageSize, setInternalPageSize] = useState(() => {
    return generatePagingOptions(pageLimits)[0].value;
  });
  const {
    paginate,
    currentPage: internalCurrentPage,
    pageData: internalData,
  } = usePagination(externallyControlled ? [] : data, {
    itemsPerPage: internalPageSize,
    autoResetPage: true,
  });

  const controlData = externallyControlled ? data : internalData;
  const controlPageSize = externallyControlled
    ? props.pageSize
    : internalPageSize;
  const controlTotalItems = externallyControlled
    ? props.totalItems
    : data.length;
  const controlCurrentPage = externallyControlled
    ? props.currentPage
    : internalCurrentPage;

  const handlePageChange = (page: number) => {
    if (externallyControlled) {
      props.onPageChange(page);
      return;
    }

    paginate(page);
  };

  const handlePageSizeChange = (pageSize: number) => {
    if (externallyControlled) {
      props.onPageSizeChange(pageSize);
      return;
    }

    paginate(1);
    setInternalPageSize(pageSize);
  };

  const clonedChildren = renderChildren(children, dataForwardProp, controlData);
  const controls = (
    <TablePaginationControls
      {...divProps}
      data={controlData}
      className={className}
      itemName={itemName}
      description={description}
      pageLimits={pageLimits}
      totalItems={controlTotalItems}
      currentPage={controlCurrentPage}
      pageSize={controlPageSize}
      onPageChange={handlePageChange}
      onPageSizeChange={handlePageSizeChange}
    />
  );

  return (
    <>
      {position === "above" && controls}
      {clonedChildren}
      {position === "below" && controls}
    </>
  );
};

export default TablePagination;
