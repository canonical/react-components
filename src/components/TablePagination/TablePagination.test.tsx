/* eslint-disable testing-library/no-node-access */
import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import TablePagination from "./TablePagination";
import userEvent from "@testing-library/user-event";

const dummyData = [
  { id: "row-1" },
  { id: "row-2" },
  { id: "row-3" },
  { id: "row-4" },
  { id: "row-5" },
];

describe("<TablePagination />", () => {
  // snapshot tests
  it("renders table pagination and matches the snapshot", () => {
    render(<TablePagination data={dummyData} />);

    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  // unit tests
  it("renders default display title correctly when no pagination takes place", () => {
    render(<TablePagination data={dummyData} />);

    expect(screen.getByRole("navigation")).toHaveTextContent(
      "Showing all 5 items"
    );
  });

  it("renders default display title correctly when pagination takes place", () => {
    render(<TablePagination data={dummyData} pageLimits={[1]} />);

    expect(screen.getByRole("navigation")).toHaveTextContent(
      "Showing 1 out of 5 items"
    );
  });

  it("has correct per page setting when changed", () => {
    render(<TablePagination data={dummyData} pageLimits={[2, 5, 10]} />);

    expect(screen.getByRole("navigation")).toHaveTextContent("2/page");
    userEvent.selectOptions(
      screen.getByRole("combobox", { name: "Items per page" }),
      "5"
    );
    expect(screen.getByRole("navigation")).toHaveTextContent("5/page");
  });

  it("resets to first page when page size is changed", () => {
    render(<TablePagination data={dummyData} pageLimits={[2, 5, 10]} />);

    expect(screen.getByRole("navigation")).toHaveTextContent("2/page");
    userEvent.selectOptions(
      screen.getByRole("combobox", { name: "Items per page" }),
      "5"
    );
    const currentPageInput = screen.getByRole("spinbutton", {
      name: "Page number",
    });
    expect(currentPageInput).toHaveValue(1);
  });

  it("should paginate correctly in incrementing or decrementing directions", async () => {
    render(<TablePagination data={dummyData} pageLimits={[1, 2, 5]} />);
    const incButton = screen.getByRole("button", { name: "Next page" });
    const decButton = screen.getByRole("button", { name: "Previous page" });
    const currentPageInput = screen.getByRole("spinbutton", {
      name: "Page number",
    });
    const pageSizeSelector = screen.getByRole("combobox", {
      name: "Items per page",
    });

    expect(currentPageInput).toHaveValue(1);
    await userEvent.click(decButton);
    expect(currentPageInput).toHaveValue(1);
    await userEvent.click(incButton);
    expect(currentPageInput).toHaveValue(2);
    await userEvent.selectOptions(pageSizeSelector, "2");
    expect(currentPageInput).toHaveValue(1);
    await fireEvent.change(currentPageInput, { target: { value: 3 } });
    expect(currentPageInput).toHaveValue(3);
    await userEvent.click(incButton);
    expect(currentPageInput).toHaveValue(3);
    await userEvent.click(decButton);
    expect(currentPageInput).toHaveValue(2);
  });
});
