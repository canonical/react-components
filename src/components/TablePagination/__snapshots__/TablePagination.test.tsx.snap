// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TablePagination /> renders table pagination and matches the snapshot 1`] = `
<div
  class="pagination"
  role="navigation"
>
  <div
    class="description"
    id="pagination-description"
  >
    Showing all 5 items
  </div>
  <button
    aria-disabled="true"
    aria-label="Previous page"
    class="p-button--base has-icon is-disabled back"
  >
    <i
      class="p-icon--chevron-down"
    />
  </button>
  <div
    class="p-form__group p-form-validation"
  >
    <label
      class="u-off-screen p-form__label"
      for="paginationPageInput"
    >
      Page number
    </label>
    <div
      class="p-form__control u-clearfix"
    >
      <input
        aria-invalid="false"
        class="p-form-validation__input u-no-margin--bottom pagination-input"
        id="paginationPageInput"
        type="number"
        value="1"
      />
    </div>
  </div>
   
  ofÂ 
  1
  <button
    aria-disabled="true"
    aria-label="Next page"
    class="p-button--base has-icon is-disabled next"
  >
    <i
      class="p-icon--chevron-down"
    />
  </button>
  <div
    class="p-form__group p-form-validation"
  >
    <label
      class="u-off-screen p-form__label"
      for="itemsPerPage"
    >
      Items per page
    </label>
    <div
      class="p-form__control u-clearfix"
    >
      <div
        class="p-form-validation__select-wrapper"
      >
        <select
          aria-describedby=""
          aria-invalid="false"
          class="p-form-validation__input u-no-margin--bottom"
          id="itemsPerPage"
        >
          <option
            value="50"
          >
            50/page
          </option>
          <option
            value="100"
          >
            100/page
          </option>
          <option
            value="200"
          >
            200/page
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;
